src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/BulkCopySendTemporalDataTypesAsStringAE.java

```diff
@@ -0,0 +1,173 @@
+@Tag(Constants.xAzureSQLDB)
+@Tag(Constants.xAzureSQLDW)
+public class BulkCopySendTemporalDataTypesAsStringAE extends AESetup {
+    static String inputFile = "BulkCopyCSVSendTemporalDataTypesAsStringForBulkCopy.csv";
+    static String encoding = "UTF-8";
+    static String delimiter = ",";
+    static String filePath = null;
+
+    static String destTableNameAE = AbstractSQLGenerator
+            .escapeIdentifier(RandomUtil.getIdentifier("sendTemporalDataTypesAsStringForBulkCopyDestTableAE"));
+
+    /**
+     * Test basic case with sendTemporalDataTypesAsStringForBulkCopy connection property, with AE enabled.
+     * 
+     * @throws SQLException
+     */
+    @Test
+    public void testSendTemporalDataTypesAsStringForBulkCopyAE() throws SQLException {
+        beforeEachSetupAE();
+        try (Connection conn = PrepUtil
+                .getConnection(AETestConnectionString + ";sendTemporalDataTypesAsStringForBulkCopy=false", AEInfo)) {
+            SQLServerBulkCSVFileRecord fileRecord = new SQLServerBulkCSVFileRecord(filePath + inputFile, encoding,
+                    delimiter, true);
+
+            testBulkCopyCSV(conn, fileRecord, destTableNameAE);
+        }
+    }
+
+    private void testBulkCopyCSV(Connection conn, SQLServerBulkCSVFileRecord fileRecord, String tableName) {
+        try (BufferedReader br = new BufferedReader(
+                new InputStreamReader(new FileInputStream(filePath + inputFile), encoding));
+                SQLServerBulkCopy bulkCopy = new SQLServerBulkCopy(conn); Statement stmt = conn.createStatement()) {
+
+            fileRecord.addColumnMetadata(1, "c1", java.sql.Types.DATE, 0, 0); // with Date
+            fileRecord.addColumnMetadata(2, "c2", java.sql.Types.TIMESTAMP, 0, 0); // with Datetime
+            fileRecord.addColumnMetadata(3, "c3", java.sql.Types.TIMESTAMP, 0, 7); // with Datetime2
+            fileRecord.addColumnMetadata(4, "c4", java.sql.Types.TIME, 0, 7); // with time
+            fileRecord.addColumnMetadata(5, "c5", microsoft.sql.Types.DATETIMEOFFSET, 0, 7); // with datetimeoffset
+            fileRecord.addColumnMetadata(6, "c6", java.sql.Types.TIMESTAMP, 0, 0); // with SmallDatetime
+            fileRecord.addColumnMetadata(7, "c7", java.sql.Types.DECIMAL, 19, 4); // with money
+            fileRecord.addColumnMetadata(8, "c8", java.sql.Types.DECIMAL, 10, 4); // with smallmoney
+
+            bulkCopy.setDestinationTableName(tableName);
+            bulkCopy.writeToServer(fileRecord);
+
+            validateValuesFromCSV(stmt, tableName, inputFile);
+        } catch (Exception e) {
+            fail(e.getMessage());
+        }
+    }
+
+    static void validateValuesFromCSV(Statement stmt, String destinationTable, String inputFile) {
+        try (BufferedReader br = new BufferedReader(
+                new InputStreamReader(new FileInputStream(filePath + inputFile), encoding));
+                ResultSet rs = stmt.executeQuery("SELECT * FROM " + destinationTable)) {

```

Comment:
This is assuming insertion order is the same as select order, which is not true. Make sure to have an ID column to be able to order the select result by ID.

Reply:
This is still not fixed.

Reply:
Why is this not fixed?

Reply:
oh never mind, you meant the row order. I'll fix that.

Reply:
I have a similar tests, you can refer to this PR: https://github.com/microsoft/mssql-jdbc/pull/1312

Reply:
Thanks, done